clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1004,6 +1014,24 @@ public ListOffsetResult() {
         return subscriptions.fetchablePartitions(tp -> !exclude.contains(tp));
     }
 
+    /**
+     * Determine which replica to read from.
+     */
+    private Node selectReadReplica(TopicPartition partition, Node leaderReplica) {
```

Comment:
Thinking about this a little more, I think we may need a tighter timeout to find a new preferred read replica. We want to give the consumer the ability to fail over quickly in the case of a failure. The fact that the metadata is exposed in a separate API makes this a bit awkward though because we have to worry about the consistency of metadata between the leader and follower. 

My basic thought is basically that we should trust the leader until our local metadata suggests that it is wrong. In other words, we should continue trying to fetch from the preferred read replica until it is marked as offline in the metadata. So if we are disconnected, we refresh metadata (as we currently do), and if the preferred read replica is still listed as live, we just try to reconnect. However, if we refresh metadata and our preferred read replica is offline, then we clear the preferred read replica and go back to the leader. Then maybe the read replica lease time could be tied to `metadata.max.age.ms` and not the request timeout and we wouldn't need any logic to update the last fetch time. Does that make any sense?

Reply:
This sounds reasonable to me. Let me make sure I understand:

* When we first set the read replica, set the timeout to `metadata.max.age.ms`, don't extend this timeout
* Upon timeout, always go back to the leader
* Also if the node is marked offline by our metadata, go back to leader